### PR TITLE
[TableGen] Simplify generated code for isSubclass

### PR DESCRIPTION
**Pull request by jayfoad**
Implement isSubclass with direct lookup into some tables instead of
nested switches.

Part of the motivation for this is improving compile time when clang-18
is used as a host compiler, since it seems to have trouble with very
large switch statements.

### CODE REVIEW

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 1]
```diff
@@ -2551,42 +2550,29 @@ static void emitIsSubclass(CodeGenTarget &Target,
...
+  OS << "  static constexpr ArrayRef<uint16_t> SuperClassTable[] = {\n";
+  OS << "    {}, // InvalidMatchClass\n";
+  OS << "    {}, // OptionalMatchClass\n";
+  for (const auto &A : Infos)
+    OS << "    " << A.Name << "_SuperClasses,\n";
+  OS << "  };\n\n";
+
+  OS << "  ArrayRef<uint16_t> SuperClasses = SuperClassTable[(unsigned)A];\n";
+  OS << "  const uint16_t *It = lower_bound(SuperClasses, (unsigned)B);\n";
+  OS << "  return It != SuperClasses.end() && *It == (unsigned)B;\n";
```
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:2575]: `binary_search`?

Comment (jayfoad) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp]: Thanks!

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 2]
```diff
@@ -2551,42 +2550,29 @@ static void emitIsSubclass(CodeGenTarget &Target,
...
-    OS << "  return false;\n";
+  OS << "  static constexpr ArrayRef<uint16_t> SuperClassTable[] = {\n";
+  OS << "    {}, // InvalidMatchClass\n";
+  OS << "    {}, // OptionalMatchClass\n";
+  for (const auto &A : Infos)
+    OS << "    " << A.Name << "_SuperClasses,\n";
+  OS << "  };\n\n";
+
+  OS << "  ArrayRef<uint16_t> SuperClasses = SuperClassTable[(unsigned)A];\n";
+  OS << "  const uint16_t *It = lower_bound(SuperClasses, (unsigned)B);\n";
```
[the remark this answers is not in the document]
Comment (jayfoad) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp]: It seems the casts were not needed anyway (at least from testing with clang 18 and gcc 13) so I removed them.

**Comment by michaelmaitland** [comment]
IIUC we could have two nested switches (constant time lookup), but now we are using binary search (logarithmic time lookup). Will this cause any time related regressions?

**Comment by jayfoad** [comment]
> IIUC we could have two nested switches (constant time lookup), but now we are using binary search (logarithmic time lookup). Will this cause any time related regressions?

Good question. I tried measuring this with `check-llvm-mc` but the run-to-run variation was pretty high. I'll see if I can detect any difference in a single execution of `llvm-mc`.

**Comment by jayfoad** [comment]
The new version uses direct lookup instead of binary search. It seems to be as fast as the nested switch version, maybe even 1% or so faster, but it is hard to tell for sure even when using `perf stat`.

The largest tables are for AArch64 and AMDGPU, both around 16 to 20 KB, but the nested switches compiled into around 30 to 40 KB of binary code so this is still a win.

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 3]
```diff
@@ -2541,53 +2541,58 @@ static void emitIsSubclass(CodeGenTarget &Target,
...
-  // If there were case statements emitted into the string stream write the
-  // default.
-  if (EmittedSwitch)
-    OS << "  }\n";
-  else
+  if (SuperClassData.empty()) {
     OS << "  return false;\n";
-
+  } else {
+    // Dump the boolean data packed into bytes.
```
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:2578]: Did you try BitVector? It could probably simplify the below code.

Comment (jayfoad) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp:2578]: Yes I looked at it but I couldn't see any way to append BitVectors, which meant I couldn't use it for SuperClassData, so I stopped looking.

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Pull request merged**